File: backend/__tests__/routes/auth.test.js
const request = require('supertest');

const app = require('../../src/app'); // Use the actual app

// --- Mock Dependencies ---
const authenticateToken = require('../../src/middleware/auth');
const AuthService = require('../../src/services/AuthService');
const logger = require('../../src/utils/logger');

// Mock the AuthService methods used by the (assumed) auth routes
jest.mock('../../src/services/AuthService', () => ({
  loginUser: jest.fn(),
  registerUser: jest.fn(),
}));

// Mock the authentication middleware
jest.mock('../../src/middleware/auth', () => jest.fn((req, res, next) => {
  const authHeader = req.headers['authorization'];
  if (authHeader && authHeader.startsWith('Bearer ')) {
    const token = authHeader.split(' ')[1];
    // Simulate token verification based on a simple mock token
    if (token === 'valid-token') {
      req.user = { id: 'mock-user-id', role: 'user' }; // Attach mock user
      return next();
    }
  }
  // Simulate failure if no token or invalid token
  return res.status(401).json({ error: 'Unauthorized: Mock Auth Failed' }); 
}));

jest.mock('../../src/utils/logger', () => ({ info: jest.fn(), error: jest.fn(), debug: jest.fn() }));

// --- Test Data ---
const mockUser = { id: 'mock-user-id', username: 'testuser', email: 'test@example.com', role: 'user' };
const mockToken = 'valid-token';
const loginCredentials = { email: 'test@example.com', password: 'password123' };
const registerData = { username: 'newuser', email: 'new@example.com', password: 'password456', role: 'user' };

// TODO: Skipping this suite as the tested routes (/api/v1/auth/login, /register, /protected)
// do not appear to exist in the current application code. AuthService is also not used.
describe.skip('Authentication API', () => {

    beforeEach(() => {
        // Reset mocks
        jest.clearAllMocks();
        AuthService.loginUser.mockReset();
        AuthService.registerUser.mockReset();
        authenticateToken.mockClear(); // Clear calls for middleware mock
    });

    // --- POST /api/v1/auth/login --- (Assuming path based on test)
    describe('POST /api/v1/auth/login', () => {
        it('should call AuthService.loginUser and return token on valid login', async () => {
            AuthService.loginUser.mockResolvedValue({ user: mockUser, token: mockToken });

            const res = await request(app)
                .post('/api/v1/auth/login') // Assuming this path exists
                .send(loginCredentials)
                .expect(200);

            expect(AuthService.loginUser).toHaveBeenCalledWith(loginCredentials.email, loginCredentials.password);
            expect(res.body).toHaveProperty('token', mockToken);
            expect(res.body).toHaveProperty('user');
            expect(res.body.user.email).toBe(loginCredentials.email);
        });

        it('should return 401 if AuthService.loginUser throws auth error', async () => {
            AuthService.loginUser.mockRejectedValue({ statusCode: 401, message: 'Invalid credentials' });

            const res = await request(app)
                .post('/api/v1/auth/login')
                .send({ email: loginCredentials.email, password: 'wrongpassword' })
                .expect(401);

            expect(AuthService.loginUser).toHaveBeenCalledWith(loginCredentials.email, 'wrongpassword');
            // Check error response based on how the actual route handler formats errors
            // This assumes the route handler catches the error and returns its message/status
            expect(res.body).toHaveProperty('error', 'Invalid credentials'); 
        });

         it('should return 500 if AuthService.loginUser throws unexpected error', async () => {
            AuthService.loginUser.mockRejectedValue(new Error('DB Error'));

            const res = await request(app)
                .post('/api/v1/auth/login')
                .send(loginCredentials)
                .expect(500); // Assuming global error handler catches and returns 500

             expect(AuthService.loginUser).toHaveBeenCalled();
             // Check error response based on global error handler
             expect(res.body).toHaveProperty('message', 'Internal Server Error'); // Or similar
        });
    });

    // --- POST /api/v1/auth/register --- (Assuming path based on test)
     describe('POST /api/v1/auth/register', () => {
        it('should call AuthService.registerUser and return success message', async () => {
            AuthService.registerUser.mockResolvedValue({ user: { ...registerData, id: 'new-user-id' } });

            const res = await request(app)
                .post('/api/v1/auth/register')
                .send(registerData)
                .expect(201);

            expect(AuthService.registerUser).toHaveBeenCalledWith(registerData);
            expect(res.body).toHaveProperty('message', 'User registered successfully');
            expect(res.body.user).toHaveProperty('email', registerData.email);
        });

        it('should return 409 if AuthService.registerUser throws conflict error', async () => {
            AuthService.registerUser.mockRejectedValue({ statusCode: 409, message: 'Username or email already exists' });

            const res = await request(app)
                .post('/api/v1/auth/register')
                .send(registerData) // Send same data again
                .expect(409);
            
            expect(AuthService.registerUser).toHaveBeenCalledWith(registerData);
            expect(res.body).toHaveProperty('error', 'Username or email already exists');
        });

         it('should return 500 if AuthService.registerUser throws unexpected error', async () => {
            AuthService.registerUser.mockRejectedValue(new Error('DB Error'));

            const res = await request(app)
                .post('/api/v1/auth/register')
                .send(registerData)
                .expect(500);

             expect(AuthService.registerUser).toHaveBeenCalled();
             expect(res.body).toHaveProperty('message', 'Internal Server Error');
        });
    });

    // --- GET /api/v1/auth/protected --- (Assuming path based on test)
    describe('GET /api/v1/auth/protected', () => {
        it('should access protected route with valid token via mock middleware', async () => {
            const res = await request(app)
                .get('/api/v1/auth/protected') // Assuming this path exists and uses authenticateToken
                .set('Authorization', `Bearer ${mockToken}`) // Use the token the mock expects
                .expect(200);

            expect(authenticateToken).toHaveBeenCalled(); // Verify middleware was called
            // Check response from the actual protected route handler (assuming it sends this message)
            expect(res.body).toHaveProperty('message', 'Access granted to protected route'); 
        });

        it('should reject protected route without token via mock middleware', async () => {
            const res = await request(app)
                .get('/api/v1/auth/protected')
                // No Authorization header
                .expect(401);

            // Middleware mock should have returned 401
            expect(res.body.error).toEqual('Unauthorized: Mock Auth Failed'); 
        });

         it('should reject protected route with invalid token via mock middleware', async () => {
            const res = await request(app)
                .get('/api/v1/auth/protected')
                .set('Authorization', `Bearer invalid-token`)
                .expect(401);

            expect(res.body.error).toEqual('Unauthorized: Mock Auth Failed');
        });
    });
});

File: backend/__tests__/routes/auth.test.py
import pytest
from src.database.models import User
from src.database.db import db
from src.routes.auth import bp as auth_bp
from flask_jwt_extended import create_access_token
import json

@pytest.fixture
def client(app):
    app.register_blueprint(auth_bp)
    return app.test_client()

def test_register_user(client):
    # Test successful registration
    response = client.post('/auth/register', json={
        'email': 'test@example.com',
        'password': 'Test1234!',
        'first_name': 'Test',
        'last_name': 'User'
    })
    assert response.status_code == 201
    data = json.loads(response.data)
    assert 'user' in data
    assert data['user']['email'] == 'test@example.com'
    
    # Test duplicate registration
    response = client.post('/auth/register', json={
        'email': 'test@example.com',
        'password': 'Test1234!'
    })
    assert response.status_code == 409

def test_login(client):
    # Create test user
    user = User(
        email='test@example.com',
        password_hash='$2b$12$EixZaYVK1fsbw1ZfbX3OXePaWxn96p36WQoeG6Lruj3vjPGga31lW',
        first_name='Test',
        last_name='User'
    )
    db.session.add(user)
    db.session.commit()
    
    # Test successful login
    response = client.post('/auth/login', json={
        'email': 'test@example.com',
        'password': 'secret'
    })
    assert response.status_code == 200
    data = json.loads(response.data)
    assert 'access_token' in data
    assert 'refresh_token' in data
    
    # Test invalid credentials
    response = client.post('/auth/login', json={
        'email': 'test@example.com',
        'password': 'wrong'
    })
    assert response.status_code == 401

def test_refresh_token(client):
    # Create test user
    user = User(
        email='test@example.com',
        password_hash='$2b$12$EixZaYVK1fsbw1ZfbX3OXePaWxn96p36WQoeG6Lruj3vjPGga31lW',
        first_name='Test',
        last_name='User'
    )
    db.session.add(user)
    db.session.commit()
    
    # Get refresh token
    refresh_token = create_access_token(identity=user.id, fresh=False)
    
    # Test token refresh
    response = client.post('/auth/refresh', headers={
        'Authorization': f'Bearer {refresh_token}'
    })
    assert response.status_code == 200
    data = json.loads(response.data)
    assert 'access_token' in data

def test_get_current_user(client):
    # Create test user
    user = User(
        email='test@example.com',
        password_hash='$2b$12$EixZaYVK1fsbw1ZfbX3OXePaWxn96p36WQoeG6Lruj3vjPGga31lW',
        first_name='Test',
        last_name='User'
    )
    db.session.add(user)
    db.session.commit()
    
    # Get access token
    access_token = create_access_token(identity=user.id)
    
    # Test current user endpoint
    response = client.get('/auth/me', headers={
        'Authorization': f'Bearer {access_token}'
    })
    assert response.status_code == 200
    data = json.loads(response.data)
    assert data['email'] == 'test@example.com'
    assert data['first_name'] == 'Test'
    assert data['last_name'] == 'User'

def test_validation_errors(client):
    # Test invalid email format
    response = client.post('/auth/register', json={
        'email': 'invalid-email',
        'password': 'Test1234!'
    })
    assert response.status_code == 400
    
    # Test short password
    response = client.post('/auth/register', json={
        'email': 'test@example.com',
        'password': 'short'
    })
    assert response.status_code == 400
    
    # Test missing fields
    response = client.post('/auth/register', json={
        'email': 'test@example.com'
    })
    assert response.status_code == 400